### PR TITLE
Manually unescape delimiter character instead of using commons-lang3

### PR DESCRIPTION
**Pull request by falaki**
Instead of depending on commons-lang3 for unescaping java strings, we could manually construct important delimiter characters from the escaped string.

### CODE REVIEW

File: src/main/scala/com/databricks/spark/csv/util/TypeCast.scala [hunk 1]
```diff
@@ -50,4 +50,31 @@ object TypeCast {
...
+       match {
+        case 't' => '\t'
+        case 'r' => '\r'
+        case 'b' => '\b'
+        case 'f' => '\f'
+        case '\"' => '\"'
+        case '\'' => '\''
+        case 'u' if str == """\u0000""" => '\u0000'
+        case '0' => '\0'
+        case _ => throw new RuntimeException(s"Unsupported special character for delimiter: $str")
```
Comment (rxin) [inline src/main/scala/com/databricks/spark/csv/util/TypeCast.scala]: IllegalArgumentException?

File: src/main/scala/com/databricks/spark/csv/util/TypeCast.scala [hunk 2]
```diff
@@ -50,4 +50,31 @@ object TypeCast {
...
+   */
+  private[csv] def toChar(str: String): Char = {
+    if (str.startsWith("\\")) {
+      str.charAt(1)
+       match {
+        case 't' => '\t'
+        case 'r' => '\r'
+        case 'b' => '\b'
+        case 'f' => '\f'
+        case '\"' => '\"'
```
Comment (rxin) [inline src/main/scala/com/databricks/spark/csv/util/TypeCast.scala]: isn't this invalid?

File: src/main/scala/com/databricks/spark/csv/util/TypeCast.scala [hunk 3]
```diff
@@ -50,4 +50,31 @@ object TypeCast {
...
+        case '\'' => '\''
+        case 'u' if str == """\u0000""" => '\u0000'
+        case '0' => '\0'
```
Comment (rxin) [inline src/main/scala/com/databricks/spark/csv/util/TypeCast.scala]: what is \0?

File: src/main/scala/com/databricks/spark/csv/util/TypeCast.scala [hunk 4]
```diff
@@ -50,4 +50,31 @@ object TypeCast {
...
   }
+
+  /**
+   * Helper method that converts string representation of a character to actual character.
+   * It handles some Java escaped strings and throws exception if given string is longer than one
+   * character.
+   *
+   */
+  private[csv] def toChar(str: String): Char = {
+    if (str.startsWith("\\")) {
```
Comment (rxin) [inline src/main/scala/com/databricks/spark/csv/util/TypeCast.scala]: use char throughout if you are using char. str.charAt(0)

File: src/test/scala/com/databricks/spark/csv/util/TypeCastSuite.scala [hunk 5]
```diff
@@ -32,4 +32,9 @@ class TypeCastSuite extends FunSuite {
       assert(TypeCast.castTo(strVal, decimalType) === new BigDecimal(decimalVal.toString))
     }
   }
+
+  test("Can parse escaped \t") {
+    val escapedStr = """\t"""
```
Comment (rxin) [inline src/test/scala/com/databricks/spark/csv/util/TypeCastSuite.scala]: if you support other chars you should add test cases for them.

File: src/main/scala/com/databricks/spark/csv/util/TypeCast.scala (hunk 2 above)
Comment (falaki) [inline src/main/scala/com/databricks/spark/csv/util/TypeCast.scala]: Depending on configured quote character, this can be a delimiter character.

Comment (rxin) [inline src/main/scala/com/databricks/spark/csv/util/TypeCast.scala]: won't user just put in a single character quote? why does it need a \

**Pull request merged**